### ISSUE TITLE
Collection changes animation support (liquid-each)

### ISSUE THREAD

**Issue by jimmyps** [ember-animation/liquid-fire#65]
The title says it all. Provides support to animate collection changes in a list, much like the one in iOS tableview. Also make sure animation in batch update is smooth enough, considering adding and removing a number of items at once. See an example here: http://www.youtube.com/watch?v=VGK_TVoPUu4

I'm betting it will require a new helper be added, like it already has for outlet, with, if, ect, but for each, i.e. {{#liquid-each items}}...{{/liquid-each}}

**Comment by runspired** [comment]
+1

**Comment by marcemira** [comment]
Any news about this?

**Comment by ef4** [comment]
For many scenarios, this is now possible. For example, this transition and template together give you animated reordering, insertion, and deletion of the list elements:

https://github.com/ef4/physical-design-demo/blob/7f9bbe0a77beb4a0e7d8a92f4f7a4472922250fb/app/transitions.js#L22-L28

https://github.com/ef4/physical-design-demo/blob/7f9bbe0a77beb4a0e7d8a92f4f7a4472922250fb/app/templates/emberconf/speakers.hbs

The downside is that it deals only with whole-list replacement. A liquid-each that observes list mutation would still be nice to add, although I would prefer to let glimmer land, because the distinction between the two cases is going to pretty much go away.

**Comment by marcemira** [comment]
:+1: Nice, let the glimmer arrives!

**Comment by g-cassie** [comment]
@ef4 We have implemented a very custom solution for animating mutating lists in our Ember product. I don't think that a simple wrapper like {{liquid-each}} could give you enough fine-grained control.  Some of the issues we ran across include:

1. Triggering animations for new/deleted list elements only after async behaviour has resolved.
2. Temporarily removing an item and then reinserting it in the same location without creating a new instance of the component.
3. Animating movement of items around a list to match the exact operation the user performed (you cannot diff two lists and know exactly what move operation was performed).
4. Animating movement of items only after async behaviour has resolved.

The approach we have used is to create a kind of shadow array to manage animations.  This shadow array will usually be in sync with the underlying data array but allows you to manipulate the data array without animating the changes immediately.  Right now it is very customized but if it's of interest to others I would be interested in developing this into an api that people can easily plug into their existing applications.

**Comment by Kuzirashi** [comment]
:+1:

**Comment by sclaxton** [comment]
@ef4  Now that glimmer has landed, are there plans to land `liquid-each`?

**Comment by ef4** [comment]
Yes, we definitely want it. No, there is no concrete timeline, it's just "as soon as somebody gets to it."

**Comment by davidgoli** [comment]
:+1:

**Comment by eric-adstage** [comment]
Just curious: has there been any more progress on this? Or has anyone forked the repo to add in this functionality? Would love to use it!

**Comment by Foko-Phil** [comment]
Hi folks. Not sure about the etiquette of this, but would it be acceptable to put a bounty on this, through Bountysource (or a preferable alternative)? This would be all sorts of useful. Thanks!

**Comment by cprussin** [comment]
Seconded.  The approach of using liquid-bind works fine when the inner elements are simple, but doing something like that when rendering more complex components for each element is way too inefficient to make a good UX.

@ef4  I forked this and looked into the liquid-each branch, and it looks really out of date.  Do you think it's worth salvaging that branch at all, or better to start fresh?

**Comment by ef4** [comment]
The liquid-each branch is probably too old to be helpful.

**Comment by Foko-Phil** [comment]
Alrighty, I've started the bounty here: https://www.bountysource.com/issues/4350805-collection-changes-animation-support-liquid-each

**Comment by Cryrivers** [comment]
i saw a branch called `liquid-each-experiment` which is updated early this year, ef4 should be working on it.

**Comment by williamweckl** [comment]
+1

**Comment by carrotalan** [comment]
+1

**Comment by ef4** [comment]
The newest work is in https://github.com/ember-animation/ember-animated

It has a working animated-each. It's not packaged up for easy configuration yet, but the demo app shows it in use on a few routes.

**Comment by denzo** [comment]
@ef4 it looks like there has been lots of amazing work done! Is there a list of things that are still need to be done for an initial release?

**Comment by ef4** [comment]
@denzo my top priority right now is generating a good quest issue so that people can all help push ember-animated over the line toward release.
